// maci-contracts/build/tasks/helpers/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EContracts = void 0;
/**
 * Enum represents deployed contracts
 */
var EContracts;
(function (EContracts) {
    EContracts["ConstantInitialVoiceCreditProxy"] = "ConstantInitialVoiceCreditProxy";
    EContracts["FreeForAllGatekeeper"] = "FreeForAllGatekeeper";
    EContracts["EASGatekeeper"] = "EASGatekeeper";
    EContracts["Verifier"] = "Verifier";
    EContracts["TopupCredit"] = "TopupCredit";
    EContracts["MACI"] = "MACI";
    EContracts["StateAq"] = "StateAq";
    EContracts["PollFactory"] = "PollFactory";
    EContracts["MessageProcessorFactory"] = "MessageProcessorFactory";
    EContracts["TallyFactory"] = "TallyFactory";
    EContracts["PoseidonT3"] = "PoseidonT3";
    EContracts["PoseidonT4"] = "PoseidonT4";
    EContracts["PoseidonT5"] = "PoseidonT5";
    EContracts["PoseidonT6"] = "PoseidonT6";
    EContracts["VkRegistry"] = "VkRegistry";
    EContracts["Poll"] = "Poll";
    EContracts["Tally"] = "Tally";
    EContracts["MessageProcessor"] = "MessageProcessor";
    EContracts["AccQueue"] = "AccQueue";
    EContracts["AccQueueQuinaryBlankSl"] = "AccQueueQuinaryBlankSl";
    EContracts["AccQueueQuinaryMaci"] = "AccQueueQuinaryMaci";
})(EContracts || (exports.EContracts = EContracts = {}));
