// maci-contracts/build/typechain-types/factories/contracts/utilities/CommonUtilities__factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CommonUtilities__factory = void 0;
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [],
        name: "VotingPeriodNotPassed",
        type: "error",
    },
];
const _bytecode = "0x6080604052348015600f57600080fd5b50603f80601d6000396000f3fe6080604052600080fdfea264697066735822122085f1886ba6769e616ea0c2faa52a8502c9a33cac0bcab8c9a9545063bf5dc1db64736f6c63430008140033";
const isSuperArgs = (xs) => xs.length > 1;
class CommonUtilities__factory extends ethers_1.ContractFactory {
    constructor(...args) {
        if (isSuperArgs(args)) {
            super(...args);
        }
        else {
            super(_abi, _bytecode, args[0]);
        }
    }
    getDeployTransaction(overrides) {
        return super.getDeployTransaction(overrides || {});
    }
    deploy(overrides) {
        return super.deploy(overrides || {});
    }
    connect(runner) {
        return super.connect(runner);
    }
    static createInterface() {
        return new ethers_1.Interface(_abi);
    }
    static connect(address, runner) {
        return new ethers_1.Contract(address, _abi, runner);
    }
}
exports.CommonUtilities__factory = CommonUtilities__factory;
CommonUtilities__factory.bytecode = _bytecode;
CommonUtilities__factory.abi = _abi;
