// maci-contracts/build/typechain-types/factories/contracts/utilities/DomainObjs__factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DomainObjs__factory = void 0;
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [],
        name: "MESSAGE_DATA_LENGTH",
        outputs: [
            {
                internalType: "uint8",
                name: "",
                type: "uint8",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
];
const _bytecode = "0x6080604052348015600f57600080fd5b5060808061001e6000396000f3fe6080604052348015600f57600080fd5b506004361060285760003560e01c8063683f3dc314602d575b600080fd5b6034600a81565b60405160ff909116815260200160405180910390f3fea2646970667358221220a0fd4f154cda9718a9e783c1d5194d3e6ac7e10bde1f415b686e640d273a817964736f6c63430008140033";
const isSuperArgs = (xs) => xs.length > 1;
class DomainObjs__factory extends ethers_1.ContractFactory {
    constructor(...args) {
        if (isSuperArgs(args)) {
            super(...args);
        }
        else {
            super(_abi, _bytecode, args[0]);
        }
    }
    getDeployTransaction(overrides) {
        return super.getDeployTransaction(overrides || {});
    }
    deploy(overrides) {
        return super.deploy(overrides || {});
    }
    connect(runner) {
        return super.connect(runner);
    }
    static createInterface() {
        return new ethers_1.Interface(_abi);
    }
    static connect(address, runner) {
        return new ethers_1.Contract(address, _abi, runner);
    }
}
exports.DomainObjs__factory = DomainObjs__factory;
DomainObjs__factory.bytecode = _bytecode;
DomainObjs__factory.abi = _abi;
