// maci-contracts/build/typechain-types/factories/contracts/gatekeepers/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SignUpTokenGatekeeper__factory = exports.SignUpGatekeeper__factory = exports.HatsGatekeeperSingle__factory = exports.HatsGatekeeperMultiple__factory = exports.HatsGatekeeperBase__factory = exports.EASGatekeeper__factory = exports.freeForAllSignUpGatekeeperSol = void 0;
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
exports.freeForAllSignUpGatekeeperSol = __importStar(require("./FreeForAllSignUpGatekeeper.sol"));
var EASGatekeeper__factory_1 = require("./EASGatekeeper__factory");
Object.defineProperty(exports, "EASGatekeeper__factory", { enumerable: true, get: function () { return EASGatekeeper__factory_1.EASGatekeeper__factory; } });
var HatsGatekeeperBase__factory_1 = require("./HatsGatekeeperBase__factory");
Object.defineProperty(exports, "HatsGatekeeperBase__factory", { enumerable: true, get: function () { return HatsGatekeeperBase__factory_1.HatsGatekeeperBase__factory; } });
var HatsGatekeeperMultiple__factory_1 = require("./HatsGatekeeperMultiple__factory");
Object.defineProperty(exports, "HatsGatekeeperMultiple__factory", { enumerable: true, get: function () { return HatsGatekeeperMultiple__factory_1.HatsGatekeeperMultiple__factory; } });
var HatsGatekeeperSingle__factory_1 = require("./HatsGatekeeperSingle__factory");
Object.defineProperty(exports, "HatsGatekeeperSingle__factory", { enumerable: true, get: function () { return HatsGatekeeperSingle__factory_1.HatsGatekeeperSingle__factory; } });
var SignUpGatekeeper__factory_1 = require("./SignUpGatekeeper__factory");
Object.defineProperty(exports, "SignUpGatekeeper__factory", { enumerable: true, get: function () { return SignUpGatekeeper__factory_1.SignUpGatekeeper__factory; } });
var SignUpTokenGatekeeper__factory_1 = require("./SignUpTokenGatekeeper__factory");
Object.defineProperty(exports, "SignUpTokenGatekeeper__factory", { enumerable: true, get: function () { return SignUpTokenGatekeeper__factory_1.SignUpTokenGatekeeper__factory; } });
