// maci-contracts/build/typechain-types/factories/contracts/gatekeepers/SignUpGatekeeper__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SignUpGatekeeper__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [
            {
                internalType: "address",
                name: "_user",
                type: "address",
            },
            {
                internalType: "bytes",
                name: "_data",
                type: "bytes",
            },
        ],
        name: "register",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "_maci",
                type: "address",
            },
        ],
        name: "setMaciInstance",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
];
class SignUpGatekeeper__factory {
    static createInterface() {
        return new ethers_1.Interface(_abi);
    }
    static connect(address, runner) {
        return new ethers_1.Contract(address, _abi, runner);
    }
}
exports.SignUpGatekeeper__factory = SignUpGatekeeper__factory;
SignUpGatekeeper__factory.abi = _abi;
