// maci-contracts/build/hardhat.config.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint-disable import/no-extraneous-dependencies */
require("@nomicfoundation/hardhat-toolbox");
const dotenv_1 = __importDefault(require("dotenv"));
require("hardhat-artifactor");
require("hardhat-contract-sizer");
require("solidity-docgen");
// Don't forget to import new tasks here
require("./tasks/deploy");
const constants_1 = require("./tasks/helpers/constants");
require("./tasks/runner/deployFull");
require("./tasks/runner/deployPoll");
require("./tasks/runner/merge");
require("./tasks/runner/prove");
require("./tasks/runner/verifyFull");
dotenv_1.default.config();
const DEFAULT_BLOCK_GAS_LIMIT = 30000000;
const DEFAULT_GAS_MUL = 2;
const TEST_MNEMONIC = "candy maple cake sugar pudding cream honey rich smooth crumble sweet treat";
const NETWORKS_RPC_URL = (0, constants_1.getNetworkRpcUrls)();
const ETHERSCAN_API_KEYS = (0, constants_1.getEtherscanApiKeys)();
const getCommonNetworkConfig = (networkName, chainId, mnemonic) => ({
    url: NETWORKS_RPC_URL[networkName],
    blockGasLimit: DEFAULT_BLOCK_GAS_LIMIT,
    gasMultiplier: DEFAULT_GAS_MUL,
    gasPrice: process.env.GAS_PRICE ? Number(process.env.GAS_PRICE) : constants_1.NETWORKS_DEFAULT_GAS[networkName],
    saveDeployments: true,
    chainId,
    accounts: {
        mnemonic: mnemonic || process.env.MNEMONIC || TEST_MNEMONIC,
        path: "m/44'/60'/0'/0",
        initialIndex: 0,
        count: 20,
    },
});
const config = {
    solidity: {
        version: "0.8.20",
        settings: {
            optimizer: {
                enabled: true,
                runs: 200,
            },
        },
    },
    defaultNetwork: "localhost",
    networks: {
        sepolia: getCommonNetworkConfig(constants_1.ESupportedChains.Sepolia, constants_1.EChainId.Sepolia),
        optimism_sepolia: getCommonNetworkConfig(constants_1.ESupportedChains.OptimismSepolia, constants_1.EChainId.OptimismSepolia),
        coverage: getCommonNetworkConfig(constants_1.ESupportedChains.Coverage, constants_1.EChainId.Coverage, TEST_MNEMONIC),
        localhost: {
            url: "http://localhost:8545",
            loggingEnabled: false,
        },
        hardhat: {
            blockGasLimit: DEFAULT_BLOCK_GAS_LIMIT,
            gasMultiplier: DEFAULT_GAS_MUL,
            gasPrice: constants_1.NETWORKS_DEFAULT_GAS[constants_1.ESupportedChains.Hardhat],
            chainId: constants_1.EChainId.Hardhat,
            accounts: {
                mnemonic: TEST_MNEMONIC,
                path: "m/44'/60'/0'/0",
                initialIndex: 0,
                count: 20,
            },
            gas: DEFAULT_BLOCK_GAS_LIMIT,
            loggingEnabled: process.env.HARDHAT_LOGGING === "true",
            allowUnlimitedContractSize: true,
            throwOnTransactionFailures: true,
            throwOnCallFailures: true,
            mining: {
                auto: true,
                interval: 100,
            },
            forking: process.env.FORKING_URL
                ? {
                    url: process.env.FORKING_URL,
                    blockNumber: process.env.FORKING_BLOCK_NUM ? parseInt(process.env.FORKING_BLOCK_NUM, 10) : 0,
                }
                : undefined,
        },
    },
    contractSizer: {
        alphaSort: true,
        runOnCompile: true,
        disambiguatePaths: false,
    },
    etherscan: {
        apiKey: {
            [constants_1.ESupportedChains.Sepolia]: ETHERSCAN_API_KEYS[constants_1.ESupportedChains.Sepolia],
            [constants_1.ESupportedChains.OptimismSepolia]: ETHERSCAN_API_KEYS[constants_1.ESupportedChains.OptimismSepolia],
        },
        customChains: [
            {
                network: constants_1.ESupportedChains.OptimismSepolia,
                chainId: constants_1.EChainId.OptimismSepolia,
                urls: {
                    apiURL: "https://api-sepolia-optimism.etherscan.io/api",
                    browserURL: "https://sepolia-optimism.etherscan.io",
                },
            },
        ],
    },
    sourcify: {
        enabled: true,
    },
    paths: {
        tests: "./tests",
        artifacts: "./artifacts",
    },
    docgen: {
        outputDir: "./docs",
        pages: "files",
        exclude: ["./trees/zeros"],
    },
    gasReporter: {
        currency: "USD",
        enabled: true,
    },
};
exports.default = config;
