// maci-contracts/build/typechain-types/factories/contracts/interfaces/IMPFactory.sol/IMessageProcessorFactory__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IMessageProcessorFactory__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [
            {
                internalType: "address",
                name: "_verifier",
                type: "address",
            },
            {
                internalType: "address",
                name: "_vkRegistry",
                type: "address",
            },
            {
                internalType: "address",
                name: "_poll",
                type: "address",
            },
            {
                internalType: "address",
                name: "_owner",
                type: "address",
            },
            {
                internalType: "enum DomainObjs.Mode",
                name: "_mode",
                type: "uint8",
            },
        ],
        name: "deploy",
        outputs: [
            {
                internalType: "address",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
];
class IMessageProcessorFactory__factory {
    static createInterface() {
        return new ethers_1.Interface(_abi);
    }
    static connect(address, runner) {
        return new ethers_1.Contract(address, _abi, runner);
    }
}
exports.IMessageProcessorFactory__factory = IMessageProcessorFactory__factory;
IMessageProcessorFactory__factory.abi = _abi;
