// maci-domainobjs/build/ts/commands/TCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TCommand = void 0;
/**
 * @notice Command for submitting a topup request
 */
class TCommand {
    /**
     * Create a new TCommand
     * @param stateIndex the state index of the user
     * @param amount the amount of voice credits
     * @param pollId the poll ID
     */
    constructor(stateIndex, amount, pollId) {
        /**
         * Create a deep clone of this TCommand
         * @returns a copy of the TCommand
         */
        this.copy = () => new TCommand(this.stateIndex, this.amount, this.pollId);
        /**
         * Check whether this command has deep equivalence to another command
         * @param command the command to compare with
         * @returns whether they are equal or not
         */
        this.equals = (command) => this.stateIndex === command.stateIndex &&
            this.amount === command.amount &&
            this.pollId === command.pollId &&
            this.cmdType === command.cmdType;
        this.cmdType = BigInt(2);
        this.stateIndex = stateIndex;
        this.amount = amount;
        this.pollId = pollId;
    }
    /**
     * Serialize into a JSON object
     */
    toJSON() {
        return {
            stateIndex: this.stateIndex.toString(),
            amount: this.amount.toString(),
            cmdType: this.cmdType.toString(),
            pollId: this.pollId.toString(),
        };
    }
    /**
     * Deserialize into a TCommand object
     * @param json - the json representation
     * @returns the TCommand instance
     */
    static fromJSON(json) {
        return new TCommand(BigInt(json.stateIndex), BigInt(json.amount), BigInt(json.pollId));
    }
}
exports.TCommand = TCommand;
