// circomkit/dist/utils/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.colors = exports.defaultConfig = exports.PRIMES = exports.PROTOCOLS = void 0;
exports.PROTOCOLS = ['groth16', 'plonk', 'fflonk'];
exports.PRIMES = ['bn128', 'bls12381', 'goldilocks', 'grumpkin', 'pallas', 'vesta', 'secq256r1'];
/** Default configurations */
exports.defaultConfig = Object.seal({
    // general settings
    protocol: 'groth16',
    prime: 'bn128',
    version: '2.1.0',
    // directories & paths
    circuits: './circuits.json',
    dirPtau: './ptau',
    dirCircuits: './circuits',
    dirInputs: './inputs',
    dirBuild: './build',
    circomPath: 'circom',
    // compiler-specific
    optimization: 1,
    inspect: true,
    include: ['./node_modules'],
    cWitness: false,
    // groth16 phase-2 settings
    groth16numContributions: 1,
    groth16askForEntropy: false,
    // solidity & calldata
    prettyCalldata: false,
    // logger
    logLevel: 'INFO',
    verbose: true,
});
/** Colors used by the logger. */
exports.colors = {
    title: '\u001b[0;34m', // blue
    success: '\u001b[0;32m', // green
    info: '\u001b[2;37m', // gray
    trace: '\u001b[2;37m', // gray
    debug: '\u001b[2;37m', // gray
    error: '\u001b[0;31m', // red
    warn: '\u001b[0;33m', // yellow
};
