// maci-contracts/build/typechain-types/factories/contracts/interfaces/IHats__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IHats__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [
            {
                internalType: "uint256",
                name: "_admin",
                type: "uint256",
            },
            {
                internalType: "string",
                name: "_details",
                type: "string",
            },
            {
                internalType: "uint32",
                name: "_maxSupply",
                type: "uint32",
            },
            {
                internalType: "address",
                name: "_eligibility",
                type: "address",
            },
            {
                internalType: "address",
                name: "_toggle",
                type: "address",
            },
            {
                internalType: "bool",
                name: "_mutable",
                type: "bool",
            },
            {
                internalType: "string",
                name: "_imageURI",
                type: "string",
            },
        ],
        name: "createHat",
        outputs: [
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "account",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "hat",
                type: "uint256",
            },
        ],
        name: "isWearerOfHat",
        outputs: [
            {
                internalType: "bool",
                name: "",
                type: "bool",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "_hatId",
                type: "uint256",
            },
            {
                internalType: "address",
                name: "_wearer",
                type: "address",
            },
        ],
        name: "mintHat",
        outputs: [
            {
                internalType: "bool",
                name: "success",
                type: "bool",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "_target",
                type: "address",
            },
            {
                internalType: "string",
                name: "_details",
                type: "string",
            },
            {
                internalType: "string",
                name: "_imageURI",
                type: "string",
            },
        ],
        name: "mintTopHat",
        outputs: [
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
];
class IHats__factory {
    static createInterface() {
        return new ethers_1.Interface(_abi);
    }
    static connect(address, runner) {
        return new ethers_1.Contract(address, _abi, runner);
    }
}
exports.IHats__factory = IHats__factory;
IHats__factory.abi = _abi;
