// maci-contracts/build/typechain-types/factories/contracts/interfaces/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IVkRegistry__factory = exports.IVerifier__factory = exports.ITallyFactory__factory = exports.IPollFactory__factory = exports.IPoll__factory = exports.IMessageProcessor__factory = exports.IMACI__factory = exports.IHats__factory = exports.IEAS__factory = exports.impFactorySol = void 0;
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
exports.impFactorySol = __importStar(require("./IMPFactory.sol"));
var IEAS__factory_1 = require("./IEAS__factory");
Object.defineProperty(exports, "IEAS__factory", { enumerable: true, get: function () { return IEAS__factory_1.IEAS__factory; } });
var IHats__factory_1 = require("./IHats__factory");
Object.defineProperty(exports, "IHats__factory", { enumerable: true, get: function () { return IHats__factory_1.IHats__factory; } });
var IMACI__factory_1 = require("./IMACI__factory");
Object.defineProperty(exports, "IMACI__factory", { enumerable: true, get: function () { return IMACI__factory_1.IMACI__factory; } });
var IMessageProcessor__factory_1 = require("./IMessageProcessor__factory");
Object.defineProperty(exports, "IMessageProcessor__factory", { enumerable: true, get: function () { return IMessageProcessor__factory_1.IMessageProcessor__factory; } });
var IPoll__factory_1 = require("./IPoll__factory");
Object.defineProperty(exports, "IPoll__factory", { enumerable: true, get: function () { return IPoll__factory_1.IPoll__factory; } });
var IPollFactory__factory_1 = require("./IPollFactory__factory");
Object.defineProperty(exports, "IPollFactory__factory", { enumerable: true, get: function () { return IPollFactory__factory_1.IPollFactory__factory; } });
var ITallyFactory__factory_1 = require("./ITallyFactory__factory");
Object.defineProperty(exports, "ITallyFactory__factory", { enumerable: true, get: function () { return ITallyFactory__factory_1.ITallyFactory__factory; } });
var IVerifier__factory_1 = require("./IVerifier__factory");
Object.defineProperty(exports, "IVerifier__factory", { enumerable: true, get: function () { return IVerifier__factory_1.IVerifier__factory; } });
var IVkRegistry__factory_1 = require("./IVkRegistry__factory");
Object.defineProperty(exports, "IVkRegistry__factory", { enumerable: true, get: function () { return IVkRegistry__factory_1.IVkRegistry__factory; } });
