// maci-contracts/build/typechain-types/factories/contracts/interfaces/IPollFactory__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IPollFactory__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [
            {
                internalType: "uint256",
                name: "_duration",
                type: "uint256",
            },
            {
                components: [
                    {
                        internalType: "uint256",
                        name: "maxMessages",
                        type: "uint256",
                    },
                    {
                        internalType: "uint256",
                        name: "maxVoteOptions",
                        type: "uint256",
                    },
                ],
                internalType: "struct Params.MaxValues",
                name: "_maxValues",
                type: "tuple",
            },
            {
                components: [
                    {
                        internalType: "uint8",
                        name: "intStateTreeDepth",
                        type: "uint8",
                    },
                    {
                        internalType: "uint8",
                        name: "messageTreeSubDepth",
                        type: "uint8",
                    },
                    {
                        internalType: "uint8",
                        name: "messageTreeDepth",
                        type: "uint8",
                    },
                    {
                        internalType: "uint8",
                        name: "voteOptionTreeDepth",
                        type: "uint8",
                    },
                ],
                internalType: "struct Params.TreeDepths",
                name: "_treeDepths",
                type: "tuple",
            },
            {
                components: [
                    {
                        internalType: "uint256",
                        name: "x",
                        type: "uint256",
                    },
                    {
                        internalType: "uint256",
                        name: "y",
                        type: "uint256",
                    },
                ],
                internalType: "struct DomainObjs.PubKey",
                name: "_coordinatorPubKey",
                type: "tuple",
            },
            {
                internalType: "address",
                name: "_maci",
                type: "address",
            },
            {
                internalType: "contract TopupCredit",
                name: "_topupCredit",
                type: "address",
            },
            {
                internalType: "address",
                name: "_pollOwner",
                type: "address",
            },
        ],
        name: "deploy",
        outputs: [
            {
                internalType: "address",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
];
class IPollFactory__factory {
    static createInterface() {
        return new ethers_1.Interface(_abi);
    }
    static connect(address, runner) {
        return new ethers_1.Contract(address, _abi, runner);
    }
}
exports.IPollFactory__factory = IPollFactory__factory;
IPollFactory__factory.abi = _abi;
