// maci-contracts/build/typechain-types/factories/contracts/interfaces/IVerifier__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IVerifier__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [
            {
                internalType: "uint256[8]",
                name: "_proof",
                type: "uint256[8]",
            },
            {
                components: [
                    {
                        components: [
                            {
                                internalType: "uint256",
                                name: "x",
                                type: "uint256",
                            },
                            {
                                internalType: "uint256",
                                name: "y",
                                type: "uint256",
                            },
                        ],
                        internalType: "struct Pairing.G1Point",
                        name: "alpha1",
                        type: "tuple",
                    },
                    {
                        components: [
                            {
                                internalType: "uint256[2]",
                                name: "x",
                                type: "uint256[2]",
                            },
                            {
                                internalType: "uint256[2]",
                                name: "y",
                                type: "uint256[2]",
                            },
                        ],
                        internalType: "struct Pairing.G2Point",
                        name: "beta2",
                        type: "tuple",
                    },
                    {
                        components: [
                            {
                                internalType: "uint256[2]",
                                name: "x",
                                type: "uint256[2]",
                            },
                            {
                                internalType: "uint256[2]",
                                name: "y",
                                type: "uint256[2]",
                            },
                        ],
                        internalType: "struct Pairing.G2Point",
                        name: "gamma2",
                        type: "tuple",
                    },
                    {
                        components: [
                            {
                                internalType: "uint256[2]",
                                name: "x",
                                type: "uint256[2]",
                            },
                            {
                                internalType: "uint256[2]",
                                name: "y",
                                type: "uint256[2]",
                            },
                        ],
                        internalType: "struct Pairing.G2Point",
                        name: "delta2",
                        type: "tuple",
                    },
                    {
                        components: [
                            {
                                internalType: "uint256",
                                name: "x",
                                type: "uint256",
                            },
                            {
                                internalType: "uint256",
                                name: "y",
                                type: "uint256",
                            },
                        ],
                        internalType: "struct Pairing.G1Point[]",
                        name: "ic",
                        type: "tuple[]",
                    },
                ],
                internalType: "struct SnarkCommon.VerifyingKey",
                name: "vk",
                type: "tuple",
            },
            {
                internalType: "uint256",
                name: "input",
                type: "uint256",
            },
        ],
        name: "verify",
        outputs: [
            {
                internalType: "bool",
                name: "",
                type: "bool",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
];
class IVerifier__factory {
    static createInterface() {
        return new ethers_1.Interface(_abi);
    }
    static connect(address, runner) {
        return new ethers_1.Contract(address, _abi, runner);
    }
}
exports.IVerifier__factory = IVerifier__factory;
IVerifier__factory.abi = _abi;
